File: app/local_plugins/translation_aliyun/android/src/main/kotlin/com/aiagent/translation_aliyun/TranslationAliyunService.kt
package com.aiagent.translation_aliyun

import android.util.Log
import com.aiagent.plugin_interface.NativeTranslationResult
import com.aiagent.plugin_interface.NativeTranslationService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.net.URLEncoder
import java.security.SecureRandom
import java.text.SimpleDateFormat
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * TranslationAliyunService — 阿里云机器翻译 API 原生实现
 *
 * 使用阿里云通用版翻译 API（mt.aliyuncs.com），HMAC-SHA1 签名。
 */
class TranslationAliyunService : NativeTranslationService {

    companion object {
        private const val TAG = "TranslationAliyunSvc"
    }

    private val client = OkHttpClient()
    private var accessKeyId: String = ""
    private var accessKeySecret: String = ""

    override fun initialize(configJson: String) {
        val cfg = JSONObject(configJson)
        val apiKey = cfg.optString("apiKey", "")
        // apiKey format: "{accessKeyId}:{accessKeySecret}"
        val parts = apiKey.split(":")
        accessKeyId = parts.getOrElse(0) { "" }
        accessKeySecret = parts.getOrElse(1) { "" }
        Log.d(TAG, "initialize: accessKeyId=${accessKeyId.take(8)}...")
    }

    override suspend fun translate(
        text: String,
        targetLang: String,
        sourceLang: String?,
    ): NativeTranslationResult = withContext(Dispatchers.IO) {
        val params = buildParams(text, targetLang, sourceLang)
        val signature = sign(params)
        params["Signature"] = signature

        val queryString = params.entries.joinToString("&") { (k, v) ->
            "${URLEncoder.encode(k, "UTF-8")}=${URLEncoder.encode(v, "UTF-8")}"
        }

        val request = Request.Builder()
            .url("https://mt.aliyuncs.com/?$queryString")
            .get()
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw Exception("Aliyun MT error: ${response.code}")
        }

        val json = JSONObject(response.body?.string() ?: "{}")
        val data = json.optJSONObject("Data")
        val translated = data?.optString("Translated", "") ?: ""

        NativeTranslationResult(
            sourceText = text,
            translatedText = translated,
            sourceLanguage = toAliyunLang(sourceLang) ?: "auto",
            targetLanguage = toAliyunLang(targetLang) ?: targetLang,
        )
    }

    override fun release() {}

    private fun buildParams(text: String, targetLang: String, sourceLang: String?): MutableMap<String, String> {
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        sdf.timeZone = TimeZone.getTimeZone("UTC")

        return mutableMapOf(
            "Action" to "TranslateGeneral",
            "Version" to "2018-10-12",
            "AccessKeyId" to accessKeyId,
            "SignatureMethod" to "HMAC-SHA1",
            "SignatureNonce" to randomNonce(),
            "SignatureVersion" to "1.0",
            "Timestamp" to sdf.format(Date()),
            "Format" to "JSON",
            "SourceLanguage" to (toAliyunLang(sourceLang) ?: "auto"),
            "TargetLanguage" to (toAliyunLang(targetLang) ?: targetLang),
            "SourceText" to text,
            "Scene" to "general",
        )
    }

    /** canonical → 阿里云机器翻译语言码（ISO 639-1 短码 + zh/zh-tw 等中文变体）。*/
    private fun toAliyunLang(code: String?): String? {
        if (code.isNullOrBlank()) return null
        return when (code.trim().uppercase()) {
            "AUTO" -> "auto"
            "ZH", "ZH-CN", "ZH-HANS" -> "zh"
            "ZH-TW", "ZH-HK", "ZH-HANT" -> "zh-tw"
            else -> code.substringBefore("-").lowercase()
        }
    }

    private fun sign(params: Map<String, String>): String {
        val sorted = params.keys.sorted()
        val canonical = sorted.joinToString("&") { k ->
            "${percentEncode(k)}=${percentEncode(params[k]!!)}"
        }
        val stringToSign = "GET&${percentEncode("/")}&${percentEncode(canonical)}"

        val mac = Mac.getInstance("HmacSHA1")
        mac.init(SecretKeySpec("$accessKeySecret&".toByteArray(Charsets.UTF_8), "HmacSHA1"))
        val rawHmac = mac.doFinal(stringToSign.toByteArray(Charsets.UTF_8))
        return Base64.getEncoder().encodeToString(rawHmac)
    }

    private fun percentEncode(value: String): String =
        URLEncoder.encode(value, "UTF-8")
            .replace("+", "%20")
            .replace("*", "%2A")
            .replace("%7E", "~")

    private fun randomNonce(): String {
        val bytes = ByteArray(16)
        SecureRandom().nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

File: app/local_plugins/vendors/translation_aliyun/ios/Classes/TranslationAliyunService.swift
import Foundation
import CommonCrypto
import ai_plugin_interface

/// Aliyun MT translation service (`mt.aliyuncs.com`, HMAC-SHA1 signed).
///
/// Config: `apiKey` formatted as `"{accessKeyId}:{accessKeySecret}"`.
public final class TranslationAliyunService: NativeTranslationService {
    private var accessKeyId: String = ""
    private var accessKeySecret: String = ""

    public init() {}

    public func initialize(configJson: String) {
        guard let data = configJson.data(using: .utf8),
              let cfg = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        let apiKey = (cfg["apiKey"] as? String) ?? ""
        let parts = apiKey.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false)
        if parts.count == 2 {
            accessKeyId = String(parts[0])
            accessKeySecret = String(parts[1])
        } else {
            accessKeyId = parts.first.map(String.init) ?? ""
            accessKeySecret = ""
        }
    }

    public func translate(
        text: String,
        targetLang: String,
        sourceLang: String?
    ) async throws -> NativeTranslationResult {
        var params = buildParams(text: text, targetLang: targetLang, sourceLang: sourceLang)
        params["Signature"] = sign(params)

        // Build the query string in the same canonical encoding the signature
        // used; otherwise the gateway rejects with `IncompleteSignature`.
        let query = params.map { k, v in
            "\(percentEncode(k))=\(percentEncode(v))"
        }.joined(separator: "&")

        guard let url = URL(string: "https://mt.aliyuncs.com/?\(query)") else {
            throw TranslationException(code: "invalid_url", message: "bad Aliyun MT URL")
        }

        let request = URLRequest(url: url)
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw TranslationException(code: "no_response", message: "no HTTP response")
        }
        if !(200..<300).contains(http.statusCode) {
            throw TranslationException(code: "http_\(http.statusCode)",
                                       message: "Aliyun MT error: \(http.statusCode)")
        }
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        let translated = ((json["Data"] as? [String: Any])?["Translated"] as? String) ?? ""

        return NativeTranslationResult(
            sourceText: text,
            translatedText: translated,
            sourceLanguage: toAliyunLang(sourceLang) ?? "auto",
            targetLanguage: toAliyunLang(targetLang) ?? targetLang
        )
    }

    public func release() {}

    // ── Helpers ───────────────────────────────────────────────────

    private func buildParams(text: String, targetLang: String, sourceLang: String?) -> [String: String] {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        formatter.timeZone = TimeZone(identifier: "UTC")

        return [
            "Action": "TranslateGeneral",
            "Version": "2018-10-12",
            "AccessKeyId": accessKeyId,
            "SignatureMethod": "HMAC-SHA1",
            "SignatureNonce": randomNonce(),
            "SignatureVersion": "1.0",
            "Timestamp": formatter.string(from: Date()),
            "Format": "JSON",
            "SourceLanguage": toAliyunLang(sourceLang) ?? "auto",
            "TargetLanguage": toAliyunLang(targetLang) ?? targetLang,
            "SourceText": text,
            "Scene": "general",
        ]
    }

    private func toAliyunLang(_ code: String?) -> String? {
        guard let raw = code?.trimmingCharacters(in: .whitespaces), !raw.isEmpty else { return nil }
        switch raw.uppercased() {
        case "AUTO":                       return "auto"
        case "ZH", "ZH-CN", "ZH-HANS":     return "zh"
        case "ZH-TW", "ZH-HK", "ZH-HANT":  return "zh-tw"
        default:
            return raw.split(separator: "-").first.map { String($0).lowercased() } ?? raw.lowercased()
        }
    }

    private func sign(_ params: [String: String]) -> String {
        let sortedKeys = params.keys.sorted()
        let canonical = sortedKeys.map { k in
            "\(percentEncode(k))=\(percentEncode(params[k] ?? ""))"
        }.joined(separator: "&")
        let stringToSign = "GET&\(percentEncode("/"))&\(percentEncode(canonical))"
        return hmacSha1Base64(stringToSign, key: "\(accessKeySecret)&")
    }

    private func percentEncode(_ value: String) -> String {
        // Aliyun signature requires RFC 3986 percent-encoding (the URL-form
        // variant used by `addingPercentEncoding(.urlQueryAllowed)` keeps `+`
        // and `*` unescaped, which breaks the signature).
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private func hmacSha1Base64(_ string: String, key: String) -> String {
        let keyData = key.data(using: .utf8) ?? Data()
        let messageData = string.data(using: .utf8) ?? Data()
        var mac = Data(count: Int(CC_SHA1_DIGEST_LENGTH))
        mac.withUnsafeMutableBytes { macBytes -> Void in
            keyData.withUnsafeBytes { keyBytes in
                messageData.withUnsafeBytes { msgBytes in
                    CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA1),
                           keyBytes.baseAddress, keyData.count,
                           msgBytes.baseAddress, messageData.count,
                           macBytes.baseAddress)
                }
            }
        }
        return mac.base64EncodedString()
    }

    private func randomNonce() -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
